Speed up code (on top of PR#199)

cippy opened the pull request:
This PR has been updated on top on the already merged PR#208 and PR#210.  Now this one mainly deals with fakes for the purpose of speeding up CardTool. The original comment on this PR is kept below for reference, although most of those items have been already included in the other PRs as said)

**--------**
This PR tries to speed up the execution of CardTool.py

- Avoid unnecessary projections , which are supposed to be done only for unfolding or multidimensional analysis
- when adding histogram members from a given group in datagroups, do not create a copy of the input histogram to return, just update the histogram that will go in output (should cause no unexpected behaviour based on some checks on systematics, especially for fakes)
- in general, whenever some operations should modify a histogram, avoid making a copy, which for big histograms is slow

These changes significantly improve the speed on the part managed by datagroups.py which prepare the boost histogram. Further improvements (presumably minor after making the sum faster) could be about summing the group members before getting to this stage, but it may not be worth for now.

The current bottleneck after them seems to be in the conversion from hist to root histograms and copying into the final root file for combinetf. This is mainly because there is a loop on many histograms, and it takes about a few seconds for each effStat boost histogram to produce and copy everything in the root file (for instance here [0]), but I haven't tested yet which is the slowest.
Maybe that part could be translated into a C++ loop somehow.

**This PR is developed on top of PR #199** 

[0] https://github.com/WMass/WRemnants/blob/main/wremnants/CardTool.py#L446-L449

davidwalter2 commented:
The impact plots from the PRValidation [1] are very different with respect to the previous merged results [2]
[1] https://cmsmwbot.web.cern.ch/WMassAnalysis/PRValidation/2023_05_02/PR204/impactsW_Wmass.png
[2] https://cmsmwbot.web.cern.ch/WMassAnalysis/PRValidation/2023_05_02/PR199/impactsW_Wmass.png
Do you know where these differences come from?

cippy commented:
Let's see, I probably had to update the branch for this PR, since at some point the previous branch for PR#199 was updated to comply with the comments (I though I had already merged from my local branch again but maybe I was still not in synch)

kdlong commented on wremnants/CardTool.py:
```diff
@@ -740,7 +742,10 @@ def writeHist(self, h, name, setZeroStatUnc=False, decorrByBin={}, hnomi=None):
                 axes.append("charge")
-            h = h.project(*axes)
+            # don't project h into itself when axes to project are all axes
+            if len(axes) < len(h.axes.name): 
```
Do you not want to explicitly check the axes?

cippy commented:
Btw, there is something fishy in the w-analysis workflow, when running the plots in the CI it finds a division by 0, I am not sure yet why it happened, but strange that it wasn't found before

cippy commented:
It turns out #PR 199 also had the decorrelated uncertainties for mT, so I am doing it again to check the impacts from CI consistently after the other fixes. If it turns out consistent I will revert the decorrelation once again, since it is probably better not to decorrelate for now

cippy commented:
So now the CI was successful (I don't know how it could fail before in the plotting part, though).
However, I don't understand why the impacts have changed so much. I am running a full local test with all stat to check impacts, they seem unnaturally large here.

I suspect some of the functions that I changed to avoid creating new boost histograms (and updating directly the input ones) might have some undesired behaviour. For the moment let's NOT merge this PR

cippy commented:
I managed to reproduce the issue with impacts, it seems to originate (but I don't know why yet) from the function to scale the luminosity here [0], but **only** when createNew is False. Maybe in that case the histograms to build the fakes are reused and scaled twice, while the copy prevents this from happening (but it is strange since the prefit plots don't seem to show any issue for fakes). Or maybe the syntax in the function is wrong, I am investigating further

Then, I modified the correlation scheme for the mT uncertainties compared to what was in PR#199 , to restore the simple single nuisances (one per charge), which is the best configuration for it.



This function is used in CI to scale to full expected luminosity (it runs on few files, and currently the scaling is of order 120), a local test with full luminosity would never be affected.

[0] https://github.com/WMass/WRemnants/blob/main/utilities/boostHistHelpers.py#L139

cippy commented:
One more thing, from my local tests with 100 files (as CI) and lumi scaling (with createNew=True, which is the working solution) I noticed that impacts are quite smaller than they would with the full statistics. This could be expected because from past studies the luminosity scaling by a large factor (120 here) was found not to fully recover the actual configuration with statistics that one is targeting: in particular, uncertainties will be more constrained, which seems to be the case indeed.

cippy commented:
Ok, the PR looks fine now, plots from CI are here:

https://cmsmwbot.web.cern.ch/WMassAnalysis/PRValidation/2023_05_05/PR204/

I think this might be ready for merge, then I'll try to understand why the function scaleToLumi with createNew=False was failing (now it uses createNew=True)

cippy commented:
Alright, now it should work, I fixed one bug for Z analyses and also added a tighter check in CardTool.checkSyst for the same-side systematics, for fakes I was getting a lot of bins where the variations were same sided, but those were also cases where Up and Down had to be equal to nominal (as for effStat in most bins) but due to numerical imprecisions they were not.

Things I am preparing for a new PR on top of this one:
- change effStat tensor removing the Up/Down axis: it seems the two are practically the mirrored image of each other wrt nominal, so we can save a good factor of 2 in memory and maybe runtime for the part which reads the boost histograms
- add more features in CardTool to set Up or Down variations equal to nominal

kdlong commented:
Thanks, can you add the before/after timing numbers here?

cippy commented:
These are some rough estimates I got. In some cases I had to run more than once to confirm, mostly I see that numbers differing by about 5-10% should be considered the same number (but it also scales a bit with the total run time).
Numbers come from running with /usr/bin/time -v,  I report the time at the line "Elapsed (wall clock) time"

Original setup: 11:28.78
Original removing dummy projections (unnecessary for the standard fit): 7:02.92 or 5:38.83 (with -v 3 or -v 4)
No histogram copy in addHist: 3:42.37
As above plus optimised fakes: 2:45.92
As above and define effStat without downUp axis (not yet in this PR): : 2:25.25

In each step there has been some additional smaller optimisation related to avoiding redundant operations. A small fraction of the time also depends on how many calls to logger are processed, or whether one was using optional scaling. For the test I tried to stick to the minimal working setup for the main analysis.

"Percent of CPU this job got" was quite stable for all tests, between 102% and 110%:
"Maximum resident set size (kbytes)" is about 16602012 (~ 16 GB) when effStat has downUp axis, otherwise it becomes ~ 9 GB

kdlong commented on utilities/boostHistHelpers.py:
```diff
@@ -129,12 +135,18 @@ def addSystAxis(h, size=1, offset=0):
-    hnew[...] = np.stack((vals, h.variances(flow=True)), axis=-1)
-    return hnew
+def clipNegativeVals(h, clipValue=0, createNew=False):
+    if createNew:
```
Can the createNew options minimally duplicate code? It's just a question of whether hfinal = h or hfinal = hnew in the end isn't it?

kdlong commented on wremnants/CardTool.py:
```diff
@@ -409,7 +409,7 @@ def checkSysts(self, var_map, proc, thresh=0.25):
                 raise e
             down_relsign = np.sign(down.values(flow=False)-hnom.values(flow=False))
-            vars_sameside = (up_relsign != 0) & (up_relsign == down_relsign)
+            vars_sameside = (up_relsign != 0) & (up_relsign == down_relsign) & np.logical_not(np.isclose(up.values(flow=False), hnom.values(flow=False), rtol=1e-07, atol=1e-08))
```
Logical not is ~ if you want this shorter (up to you)

cippy commented:
This PR has to be updated once PR#208 is merged (which is a subset of this one, to facilitate review).
After that, this PR will just feature the changes for the fakes, to make running on them faster

davidwalter2 commented on wremnants/datasets/datagroups.py:
```diff
@@ -220,13 +246,41 @@ def setHists(self, baseName, syst, procsToRead=None, label=None, nominalIfMissin
+                        # apply the correct scale for fakes
+                        scaleProcForFake = self.groups[nameFake].scale(member)
+                        logger.debug(f"Summing hist {read_syst} for {member.name} to {nameFake} with scale = {scaleProcForFake}")
+                        hProcForFake = scaleProcForFake * h.copy()
```
Can we even avoid this h.copy by writing in the next line:
```
histForFake = hh.addHists(hProcForFake, histForFake, scale1=scaleProcForFake, createNew=False) if histForFake else hProcForFake.copy()
```
and then in hh.addHists we add the optional argument to scale the first histogram?

davidwalter2 reviewed (commented):
Apart from the two comments I made the PR looks good to me

cippy commented:
Can I merge this one? I would do it, but I prefer to wait until everyone agrees
